File: src/main/java/com/example/jwttest/jwt/JwtFilter.java
package com.example.jwttest.jwt;

import com.example.jwttest.exception.CustomException;
import io.jsonwebtoken.ExpiredJwtException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class JwtFilter extends OncePerRequestFilter {

    private static final Logger log = LoggerFactory.getLogger(JwtFilter.class);
    public static final String address = "http://localhost:8080";
    private final TokenProvider tokenProvider;
    private final RestTemplate restTemplate;

    public JwtFilter(TokenProvider tokenProvider, RestTemplate restTemplate) {
        this.tokenProvider = tokenProvider;
        this.restTemplate = restTemplate;
    }

    // TODO refresh token 은 나중에 따로 받는걸로?
    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        String URI = request.getRequestURI();
        log.info("URI : " + URI);

        String accessToken = resolveAccessToken(request);                   // 토큰 정보 꺼내오고,
        if(accessToken == null) {                                           // 토큰 없으면 필터 동작 끝내기
            filterChain.doFilter(request, response);
            return;
        }

        try {
             String newAccessToken = tokenProvider.validateToken(accessToken);
        } catch (ExpiredJwtException e) {
            log.error("Access Token expired");
            // TODO 실제로는 앱/웹 단에 요청을 해야 하는데, 해당 api 를 받은 앱/웹은 서버에 refreshToken 을 전달해야 한다. 여기서는 Mock Controller 로 보낸다.
            String refreshToken = restTemplate.getForObject(address + "/api/mock/getRefresh" + "?originURI=" + URI + "&userId=" + e.getClaims().getSubject(), String.class);

            // TODO 여기서 로직을 어떻게 처리할건지?

        } catch (CustomException e) {
            log.error("Token is not valid : " + e.getMessage());
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, e.getMessage());
        }

    }


    /**
     * 헤더에서 accessToken, refreshToken 을 가져온다.
     */
    private String resolveAccessToken(HttpServletRequest request) {
        String bearerToken = request.getHeader("Authorization");
        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
            return bearerToken.substring(7);
        }
        return null;
    }
}

File: src/main/kotlin/com/example/jwttest/jwt/JwtFilter.kt
package com.example.jwttest.jwt

import com.example.jwttest.dto.TokenDto
import com.example.jwttest.exception.CustomException
import org.slf4j.LoggerFactory
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.util.StringUtils
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 토큰의 인증정보를 SecurityContext 에 보내는 역할
 * OncePerRequestFilter : 모든 요청에 대해 단 한번만 필터링을 수행한다. (기본 서블릿 필터는 DispatcherServlet 을 들어가고 나올 때마다 수행된다.)
 */
class JwtFilter(
        private val tokenProvider: TokenProvider,
        private val jwtAuthenticationEntryPoint: JwtAuthenticationEntryPoint
) : OncePerRequestFilter() {

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        httpServletRequest: HttpServletRequest,
        httpServletResponse: HttpServletResponse,
        filterChain: FilterChain
    ) {
        log.debug("URI : " + httpServletRequest.requestURI)
        val tokenDto = resolveToken(httpServletRequest)                   // 토큰 정보를 꺼내온다.
        if (!tokenDto.isPresent) {
            filterChain.doFilter(httpServletRequest, httpServletResponse) // 토큰이 없으면 토큰 검증을 실시하지 않는다.
            return
        }
        try {
            val newTokens = tokenProvider.validateToken(tokenDto.get().token, tokenDto.get().refreshToken)
            processValidTokens(tokenDto.get().token, newTokens, httpServletRequest, httpServletResponse, filterChain)
        } catch (e: CustomException) {
            jwtAuthenticationEntryPoint.commence(httpServletRequest, httpServletResponse, e)
        }
    }

    /**
     * Request Header 에서 토큰 정보를 꺼내오기 위한 메서드
     */
    private fun resolveToken(request: HttpServletRequest): Optional<TokenDto> {
        var accessToken = request.getHeader(AUTHORIZATION_HEADER)                                     // 헤더에서 토큰 정보를 꺼내온다.
        val refreshToken = request.getHeader(REFRESH_TOKEN_HEADER)
        if (StringUtils.hasText(accessToken) && accessToken.startsWith("Bearer ")) {            // 토큰 정보가 존재하고, Bearer로 시작하는 경우 (RFC 6570)
            accessToken = accessToken.split(" ".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()[1].trim { it <= ' ' }                                                 // Bearer 다음 문자열을 반환한다.
            return Optional.of(TokenDto(accessToken, refreshToken))
        }
        return Optional.empty()
    }

    /**
     * 토큰이 유효할 때 SecurityContext 에 저장하는 메서드
     */
    @Throws(IOException::class, ServletException::class)
    private fun processValidTokens(
        accessToken: String?,
        newTokens: TokenDto?,
        request: HttpServletRequest,
        response: HttpServletResponse,
        chain: FilterChain
    ) {
        if (StringUtils.hasText(newTokens?.token)) {
            val authentication = tokenProvider.getAuthentication(newTokens?.token)
            SecurityContextHolder.getContext().authentication = authentication
            log.debug(authentication.name + " 의 인가 허용")
            if (newTokens?.token != accessToken) {                                              // accessToken 이 만료되어서 재발급된 경우
                log.debug("new Access token 발급해 Security Context 에 '{}' 인증 정보를 저장. 인증 정보 : {}", authentication.name, authentication.authorities)
                response.addHeader(AUTHORIZATION_HEADER, "Bearer " + newTokens?.token)
                response.addHeader(REFRESH_TOKEN_HEADER, newTokens?.refreshToken)
            }
        } else {
            log.debug("유효한 토큰이 없습니다. Security Context 에 저장하지 않음")
        }
        chain.doFilter(request, response)
    }

    companion object {
        private val log = LoggerFactory.getLogger(JwtFilter::class.java)
        const val AUTHORIZATION_HEADER = "Authorization" // 헤더의 키 값
        const val REFRESH_TOKEN_HEADER = "refreshToken" // refresh token 키 값
    }
}